demos/coreMQTT_Agent/mqtt_agent_task.c

```diff
@@ -0,0 +1,1025 @@
+                                       RETRY_BACKOFF_BASE_MS,
+                                       RETRY_MAX_BACKOFF_DELAY_MS,
+                                       RETRY_MAX_ATTEMPTS );
+
+    /* Attempt to connect to MQTT broker. If connection fails, retry after a
+     * timeout. Timeout value will exponentially increase until the maximum
+     * number of attempts are reached.
+     */
+    do
+    {
+        /* Establish a TCP connection with the MQTT broker. This example connects to
+         * the MQTT broker as specified in democonfigMQTT_BROKER_ENDPOINT and
+         * democonfigMQTT_BROKER_PORT at the top of this file. */
+        LogInfo( ( "Creating a TLS connection to %s:%d.",
+                   democonfigMQTT_BROKER_ENDPOINT,
+                   democonfigMQTT_BROKER_PORT ) );
+
+        xNetworkStatus = SecureSocketsTransport_Connect( pxNetworkContext, &xServerInfo, &xSocketConfig );
+
+        xConnected = ( xNetworkStatus == TRANSPORT_SOCKET_STATUS_SUCCESS ) ? pdPASS : pdFAIL;
+
+        if( !xConnected )
+        {
+            LogWarn( ( "Connection to the broker failed. Attempting connection retry after backoff delay." ) );
+
+            /* As the connection attempt failed, we will retry the connection after an
+             * exponential backoff with jitter delay. */
+
+            /* Calculate the backoff period for the next retry attempt and perform the wait operation. */
+            xBackoffStatus = prvBackoffForRetry( &xReconnectParams );
+        }
+    } while( ( xConnected != pdPASS ) && ( xBackoffStatus == pdPASS ) );
+
+    /* Set the socket wakeup callback and ensure the read block time. */
+    if( xConnected )
+    {
+        SOCKETS_SetSockOpt( pxNetworkContext->pParams->tcpSocket,
+                            0,
+                            SOCKETS_SO_RCVTIMEO,
+                            &xTransportTimeout,
+                            sizeof( TickType_t ) );
+    }
+
+    return xConnected;
+}
+
+/*-----------------------------------------------------------*/
+
+static BaseType_t prvSocketDisconnect( NetworkContext_t * pxNetworkContext )
+{
+    LogInfo( ( "Disconnecting TLS connection.\n" ) );
+    TransportSocketStatus_t xNetworkStatus = SecureSocketsTransport_Disconnect( pxNetworkContext );
+
+    return ( xNetworkStatus == TRANSPORT_SOCKET_STATUS_SUCCESS ) ? pdPASS : pdFAIL;
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvIncomingPublishCallback( MQTTAgentContext_t * pMqttAgentContext,
+                                        uint16_t packetId,
+                                        MQTTPublishInfo_t * pxPublishInfo )
+{
+    bool xPublishHandled = false;
+    char cOriginalChar, * pcLocation;
+
+    ( void ) packetId;
+
+    /* Fan out the incoming publishes to the callbacks registered using
+     * subscription manager. */
+    xPublishHandled = handleIncomingPublishes( ( SubscriptionElement_t * ) pMqttAgentContext->pIncomingCallbackContext,
```

Comment:
Would have been more readable if the name of the function has the `MQTTSubscriptionManager_` prefix in it to clearly mention that this publish callback delegates the incoming publish to the subscription manager
```suggestion
    xPublishHandled = handleIncomingPublishes( ( SubscriptionElement_t * ) pMqttAgentContext->pIncomingCallbackContext,
```

Reply:
I think the `MQTTSubscriptionManager_` prefix sort of implies it's a separate library instead of part of the demo, so that might be confusing. As it is now, the subscription manager code is the same across all the repositories this demo is in, and I don't want to change it in only this location.